### ISSUE TITLE
JVM crashes when debugging with PyCharm  on Windows

### ISSUE THREAD

**Issue by fbl100** [jpype-project/jpype#346]
When stepping through code, the first call to a java function causes the python script to crash when using PyCharm.  When running (not debugging) it works fine.  PyCharm debugging works fine on the mac and as far as I can tell, everything else works on windows.  I have tried building from source on windows and cannot build.  Any support would be appreciated.  I am also willing to help contribute if necessary but I don't know exactly where to start, especially when it comes to building from scratch.  I'm a pretty competent java dev with C/C++ knowledge.

Is this a bug with the most recent release (0.6.1)?  If so, maybe this is a bug against building from scratch on Windows (which is what I did for my mac)

**Comment by Thrameos** [comment]
I have not seen any issues with PyCharm with the latest when I tried, but I did not run the debugging mode.  However, it is not my main work environment.  I primarily test with Spyder and Jupyter.   

There is a serious issue with the current release regarding unattached threads.  If more than one thread attempts to access a resource from java then there will be a segfault because the JVM does not recognize the origination threat (unattached).  If the thread is a attached then everything works properly.  Unfortunately, some GUIs like PyCharm use a second thread for purposes of debugging or to interrogate the class (for example getting a tab completion list), and when those threads attempt to access the java resource they will do so as an unattached thread.   Thus the problem is not using the module as intended but rather the access through other structures.  There wasn't much hardening on the unattached threads though so when it happens it just went straight to a segfault.

I know the latest development version has a fix the issue unattached threads.  I believe it is also in the current master, but it was a few months ago that I worked on that issue so I can't recall the if that was a full or partial fix.

**Comment by fbl100** [comment]
It works in PyCharm, just not in debugging mode.  On my mac, I was able to build per the instructions, but I'm having a heck of a time building on Windows.  I can provide build logs and procedures if you'd like.  Once we get it worked out (building from source on windows), I'd be happy to help update the documentation.  Just let me know what you'd like me to provide.

**Comment by Thrameos** [comment]
Okay I can look over the logs.  But in the meantime here is a dump of what I see when I build on my test machine. 

Building with current master, Python3.6 (miniconda install), Oracle jdk 1.8.0_144, Netbeans 8.2, and Visual studio tools for Python 3.  I am using cygwin for the shell, but there should be no difference between that and running with the Visual studio command shell for the 64 bit command line compiler.  The last line fails because it does not have the rights to alter the site packages but the rest of the build is successful.  The only argument that is required is pointing it to the ant on the system as I am using "ant.bat" included in netbeans rather than "ant.exe" from the standalone version.

[f.log](https://github.com/jpype-project/jpype/files/2435483/f.log)

After the build I would just run ant on the test directory and then call nose tests to make sure everything was properly installed.   Both of these steps should not result in any problems.

Running the ant build of the testsuite.
```
$ ant -f test/build.xml
Buildfile: F:\nelson85pl\nelson85\devel\open\jpype\test\build.xml

compile:
    [javac] F:\nelson85pl\nelson85\devel\open\jpype\test\build.xml:15: warning: 'includeantruntime' was not set, defaulting to build.sysclasspath=last; set to false for repeatable builds
    [javac] Compiling 28 source files to F:\nelson85pl\nelson85\devel\open\jpype\test\classes
    [javac] warning: [options] bootstrap class path not set in conjunction with -source 1.7
    [javac] 1 warning

test:

BUILD SUCCESSFUL
Total time: 1 second
```

Executing the test suite. (with nose and mock modules installed)
```
$ python3 -m nose test/jpypetest
...........................................................................................S.........S.SSSS..............S...........................................................
----------------------------------------------------------------------
Ran 181 tests in 3.043s

OK (SKIP=7)
```

I hope this points you in the right direction as it should pretty much be the same as the documentation.

**Comment by fbl100** [comment]
That actually helped... I was able to get it to install.  Even though ant.bat is on my path I still had to use --ant=ant.bat to get it to work.
  When I run pip list, I see v0.6.3 installed, but I get an error when starting the jvm:

    startJVM( getDefaultJVMPath(), "-ea", "-Djava.class.path=%s" % classpath)
  File "C:\Users\frank\AppData\Local\Programs\Python\Python36\lib\site-packages\jpype1-0.6.3-py3.6-win-amd64.egg\jpype\_core.py", line 70, in startJVM
    _jpype.startup(jvm, tuple(args), True)
RuntimeError: Unable to load DLL [C:\Program Files\Java\jdk1.8.0_181\jre\bin\server\jvm.dll], error = The specified module could not be found.

        at unknown(native\common\include\jp_platform_win32.h:58)

I have verified that the jvm.dll does in-fact exist.

FWIW I'm running on a clean Windows 10 Home virtualbox with only the stuff required to get this to work installed (vstools, java, ant, git, and Python 3.6)

**Comment by Thrameos** [comment]
The `ant.bat` issue is that python setup tools does not appear to search for bat files thus you must explicitly specify ant if you use the netbeans copy; it works out of the box is `ant.exe` is in the path.  It should be possible to fix in the setupext, but I haven't found a pattern for it yet.

 As far as the jvm.dll, are you sure that the python build and the jvm have matching architectures?  (We have a check for it but i want to make sure it is not that issue.)   Are both the jvm and python 64 bit compiled?   (given the amd64.egg it seems unlikely to be this problem.)

```
C:\Users\nelson85>"C:\ProgramData\Miniconda3\python.exe"
Python 3.6.3 |Anaconda, Inc.| (default, Oct 15 2017, 03:27:45) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import jpype
>>> jpype.startJVM(jpype.getDefaultJVMPath())
>>> print(jpype.getDefaultJVMPath())
C:\Program Files\Java\jre1.8.0_181\bin\server\jvm.dll
>>> import platform
>>> platform.architecture()
('64bit', 'WindowsPE')
>>>
```

Second, it is possible that the jvm is not set to be executable on your system?  (Should be read & execute privilege.)  It is pretty hard to trigger this one as one has to try to remove the privilege, but it can happen if the directory was copied rather than installed.

So we can see if it exists...
```
C:\Users\nelson85>dir "C:\Program Files\Java\jre1.8.0_181\bin\server\jvm.dll"
 Volume in drive C has no label.
 Volume Serial Number is B84B-AEE8

 Directory of C:\Program Files\Java\jre1.8.0_181\bin\server

08/18/2018  06:19 AM         8,835,448 jvm.dll
               1 File(s)      8,835,448 bytes
               0 Dir(s)  218,960,539,648 bytes free
```

And then see if it executable...
```
C:\Program Files\Java\jre1.8.0_181\bin\server\jvm.dll NT AUTHORITY\SYSTEM:(I)(F)

                                                      BUILTIN\Administrators:(I)
(F)
                                                      BUILTIN\Users:(I)(RX)

Successfully processed 1 files; Failed processing 0 files
```

On my system I am using 1.8.0_181 as well with no issues.
```
$ md5sum "C:\Program Files\Java\jre1.8.0_181\bin\server\jvm.dll"
\f71a984652579a868003b82d318d0515 *C:\\Program Files\\Java\\jre1.8.0_181\\bin\\server\\jvm.dll
```

The next possible problem would be that the dll is properly in place, but some required dll that it uses is not in place or corrupt.   Here is what I see on my system.

```
$ cygcheck "C:\Program Files\Java\jre1.8.0_181\bin\server\jvm.dll"
C:\Program Files\Java\jre1.8.0_181\bin\server\jvm.dll
  C:\Windows\system32\KERNEL32.dll
    C:\Windows\system32\API-MS-Win-Core-RtlSupport-L1-1-0.dll
    C:\Windows\system32\ntdll.dll
    C:\Windows\system32\KERNELBASE.dll
    C:\Windows\system32\API-MS-Win-Core-ProcessThreads-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Heap-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Memory-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Handle-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Synch-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-File-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-IO-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-ThreadPool-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-LibraryLoader-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-NamedPipe-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Misc-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-SysInfo-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Localization-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-ProcessEnvironment-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-String-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Debug-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-ErrorHandling-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Fibers-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Util-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Core-Profile-L1-1-0.dll
    C:\Windows\system32\API-MS-Win-Security-Base-L1-1-0.dll
  C:\Windows\system32\USER32.dll
    C:\Windows\system32\GDI32.dll
      C:\Windows\system32\API-MS-Win-Core-LocalRegistry-L1-1-0.dll
      C:\Windows\system32\LPK.dll
        C:\Windows\system32\USP10.dll
          C:\Windows\system32\msvcrt.dll
            C:\Windows\system32\API-MS-Win-Core-Console-L1-1-0.dll
            C:\Windows\system32\API-MS-Win-Core-DateTime-L1-1-0.dll
  C:\Windows\system32\ADVAPI32.dll
    C:\Windows\system32\API-MS-WIN-Service-Core-L1-1-0.dll
    C:\Windows\system32\API-MS-WIN-Service-winsvc-L1-1-0.dll
    C:\Windows\system32\API-MS-WIN-Service-Management-L1-1-0.dll
    C:\Windows\system32\API-MS-WIN-Service-Management-L2-1-0.dll
    C:\Windows\system32\RPCRT4.dll
      C:\Windows\system32\API-MS-Win-Core-Interlocked-L1-1-0.dll
      C:\Windows\system32\API-MS-Win-Core-DelayLoad-L1-1-0.dll
  C:\Windows\system32\WSOCK32.dll
    C:\Windows\system32\WS2_32.dll
      C:\Windows\system32\NSI.dll
  C:\Windows\system32\WINMM.dll
  C:\Windows\system32\VERSION.dll
  C:\Windows\system32\PSAPI.DLL
  C:\Windows\system32\MSVCR100.dll
```

Seem pretty unlikely as all of those dlls should be system ones.  

Do you have another JVM installed that you can point the `JAVA_HOME` to?

I am slightly perplexed as I had a group of 5 students install and run jpype this summer on Windows 10 systems without issues other than one student who had problems on the build (installed wrong python/jvm arch, installed incorrect visual studio, tried to build without Visual studio path, and didn't have ant).   But given that you have gotten through the build process I don't know much that can go wrong, so I am trying make some educated(?) guesses.

**Comment by fbl100** [comment]
Success!  It turns out that I didn't have the Visual C++ 2010 runtime installed, only 2015.  cygcheck did the trick.  I would think that when I installed the JDK or JRE that would have come along for the ride, but I guess not.  Regardless, I can now build and run.  I can also confirm that the version that gets installed with pip crashes when you debug in PyCharm.  When I build from master, the PyCharm problem goes away.

To wrap this up, I'm going to do another fresh install on a clean Windows 10 VM and document exactly what I needed to install and the steps necessary.  I think it's a little more complicated than the current docs suggest, especially if you don't already have ant (or Netbeans), and VisualStudio installed.  When I'm done, I'd be happy to contribute that back to the docs.

Thanks for your help - JPype is enabling our engineers (who use python) to access significant functionality that would have previously required them to learn java.

**Comment by fbl100** [comment]
To close this out, this is the minimal set of pre-requisites that I needed to install to get it to build/run on a clean Windows VM (jdk8 and python are assumed to be there already):

Microsoft Visual Studio 2010 Redistributable Package (x64)
https://www.microsoft.com/en-us/download/details.aspx?id=14632

Apache Ant (tested using 1.9.13)
https://ant.apache.org/bindownload.cgi

Microsoft Build Tools 2015 Update 3
https://visualstudio.microsoft.com/vs/older-downloads/

The Build Tools 2015 is a pain to find.  Microsoft really wants people to download the latest version.  To get to it from the above URL, click on "Redistributables and Build Tools", then select Microsoft Build Tools 2015 Update 3.

Finally, the setup.py must be run from the VS2015 x64 developer command prompt.

**Comment by Thrameos** [comment]
Excellent.  Do you feel that a patch to the current documentation is required?  And if so can you submit it?

**Comment by Thrameos** [comment]
Addressed in #363.
